File: q0496/q0496_forward.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        // get the next great elements of nums2 into a map
        Map<Integer, Integer> map = new HashMap<>();
        Deque<Integer> stack = new ArrayDeque<>();

        // build the monotonic increasing stack through looping forward
        for (int currNum : nums2) {
            // currNum will the next greater element for all the elements on the stack that
            // is currently smaller than it
            while (!stack.isEmpty() && currNum > stack.peek()) {
                int prevNum = stack.pop();
                map.put(prevNum, currNum);
            }
            stack.push(currNum);
        }

        // collect the results from the map
        int[] res = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            res[i] = map.getOrDefault(nums1[i], -1);
        }
        return res;
    }
}

File: q0496/q0496_forward.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        monoStack = []
        nextGreaterElementMap = {}
        for num in nums2:
            # all the numbers on the monotonically increasing stack
            # if they are less than the current num
            # their next greater element in nums2 is current num
            while len(monoStack) != 0 and num > monoStack[-1]:
                prevNum = monoStack.pop()
                nextGreaterElementMap[prevNum] = num
            
            # keep track of the current num
            monoStack.append(num)
        
        res = [-1] * len(nums1)
        for i in range(len(nums1)): 
            if nums1[i] in nextGreaterElementMap: 
                res[i] = nextGreaterElementMap[nums1[i]]
        return res
